File: speckit-chat-plugin/src/main/kotlin/com/speckit/plugin/model/PipelineModels.kt
package com.speckit.plugin.model

import java.io.File

enum class StepStatus { NOT_STARTED, READY, IN_PROGRESS, COMPLETED, BLOCKED }

data class ArtifactCheck(
    val relativePath: String,
    val label: String,
    val isDirectory: Boolean = false,
    val requireNonEmpty: Boolean = false,
    val isRepoRelative: Boolean = false,
    val altRelativePath: String? = null
)

data class CheckResult(val artifact: ArtifactCheck, val exists: Boolean, val detail: String, val resolvedFile: File? = null)

data class ChecklistStatus(
    val fileName: String,
    val total: Int,
    val completed: Int,
    val file: File
) {
    val isComplete get() = completed >= total
}

data class PipelineStepDef(
    val number: Int,
    val id: String,
    val name: String,
    val description: String,
    val isOptional: Boolean,
    val prerequisites: List<ArtifactCheck>,
    val outputs: List<ArtifactCheck>,
    val handsOffTo: List<String>,
    val agentFileName: String,
    val parentId: String? = null,
    val defaultArgs: String = ""
)

class PipelineStepState {
    @Volatile var status: StepStatus = StepStatus.NOT_STARTED
    var prerequisiteResults: List<CheckResult> = emptyList()
    var outputResults: List<CheckResult> = emptyList()
}

data class FeaturePaths(
    val basePath: String,
    val branch: String,
    val isFeatureBranch: Boolean,
    val featureDir: String?
)

data class FeatureEntry(
    val dirName: String,
    val path: String,
    var completedSteps: Int = 0,
    var totalOutputSteps: Int = 0
)

File: speckit-chat-vscode/src/model/pipeline-models.ts
export enum StepStatus {
    NOT_STARTED = 'NOT_STARTED',
    READY = 'READY',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETED = 'COMPLETED',
    BLOCKED = 'BLOCKED',
}

export interface ArtifactCheck {
    relativePath: string;
    label: string;
    isDirectory?: boolean;
    requireNonEmpty?: boolean;
    isRepoRelative?: boolean;
    altRelativePath?: string;
}

export interface CheckResult {
    artifact: ArtifactCheck;
    exists: boolean;
    detail: string;
    resolvedFile?: string;
}

export interface ChecklistStatus {
    fileName: string;
    total: number;
    completed: number;
    filePath: string;
    isComplete: boolean;
}

export interface PipelineStepDef {
    number: number;
    id: string;
    name: string;
    description: string;
    isOptional: boolean;
    prerequisites: ArtifactCheck[];
    outputs: ArtifactCheck[];
    handsOffTo: string[];
    agentFileName: string;
    parentId?: string;
    defaultArgs: string;
}

export class PipelineStepState {
    status: StepStatus = StepStatus.NOT_STARTED;
    prerequisiteResults: CheckResult[] = [];
    outputResults: CheckResult[] = [];
}

export interface FeaturePaths {
    basePath: string;
    branch: string;
    isFeatureBranch: boolean;
    featureDir: string | undefined;
}

export interface FeatureEntry {
    dirName: string;
    path: string;
    completedSteps: number;
    totalOutputSteps: number;
}
